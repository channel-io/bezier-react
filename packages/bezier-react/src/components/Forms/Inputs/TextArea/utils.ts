import { type SemanticNames } from '~/src/foundation'

interface CalculateBgColorProps {
  focused: boolean
  hasError: boolean
  readOnly: boolean
}

export function getTextAreaBgColorSemanticName({
  focused,
  hasError,
  readOnly,
}: CalculateBgColorProps): SemanticNames {
  if (readOnly) { return 'bg-grey-lighter' }
  if (focused || hasError) { return 'bg-white-normal' }
  return 'bg-grey-lightest'
}
