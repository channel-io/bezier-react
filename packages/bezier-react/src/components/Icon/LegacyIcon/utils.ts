/* Internal Dependencies */
// eslint-disable-next-line no-restricted-imports
import icons, { type IconName } from '../generated'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { Icon, AllIcon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-react'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(<svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
