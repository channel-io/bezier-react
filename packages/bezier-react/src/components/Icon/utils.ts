import { type SVGProps } from 'react'

const BEZIER_ICON_SYMBOL = Symbol('bezier.icon')

export type IconSource = (props: SVGProps<SVGSVGElement>) => JSX.Element

export type BezierIcon = IconSource & { [BEZIER_ICON_SYMBOL]: true }

export function isBezierIcon(arg: unknown): arg is BezierIcon {
  return (
    typeof arg === 'function' &&
    arg[BEZIER_ICON_SYMBOL] === true
  )
}

export function createBezierIcon(source: IconSource): BezierIcon {
  source[BEZIER_ICON_SYMBOL] = true
  return source as BezierIcon
}
