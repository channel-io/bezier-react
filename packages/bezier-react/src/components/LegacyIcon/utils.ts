import {
  type IconName,
  icons,
} from '@channel.io/bezier-icons'

/**
 * @deprecated Please import and use individual icons.
 * @example
 * import { AllIcon } from '@channel.io/bezier-icons'
 * import { Icon, type IconProps } from '@channel.io/bezier-react'
 * <Icon source={AllIcon} color="bg-black-dark" />
 * @example <caption>How to validate the bezier icon source</caption>
 * import { isBezierIcon, AllIcon } from '@channel.io/bezier-icons'
 * isBezierIcon(AllIcon) // true
 * isBezierIcon(() => <svg>...</svg>) // false
 * @example <caption>Legacy icon component is still available. but it will be removed in future versions!</caption>
 * import { LegacyIcon, type LegacyIconProps } from '@channel.io/bezier-react'
 * <LegacyIcon name="all" color="bg-black-dark" />
 */
export const isIconName = (name: any): name is IconName => Object.keys(icons).includes(name)
