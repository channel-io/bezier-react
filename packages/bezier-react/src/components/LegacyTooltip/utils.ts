import {
  type GetReplacement,
  type GetTooltipStyle,
  LegacyTooltipPosition,
} from './LegacyTooltip.types'

export function getReplacement({
  tooltip,
  keepInContainer,
  placement,
  rootElement,
}: GetReplacement): LegacyTooltipPosition {
  if (!keepInContainer) {
    return placement
  }

  const {
    width: tooltipWidth,
    height: tooltipHeight,
    top: tooltipTop,
    left: tooltipLeft,
  } = tooltip.getBoundingClientRect()

  const {
    width: rootWidth,
    height: rootHeight,
    top: rootTop,
    left: rootLeft,
  } = rootElement.getBoundingClientRect()

  const isOverTop = tooltipTop < rootTop
  const isOverBottom = tooltipTop + tooltipHeight > rootTop + rootHeight
  const isOverLeft = tooltipLeft < rootLeft
  const isOverRight = tooltipLeft + tooltipWidth > rootLeft + rootWidth

  if (isOverTop) {
    if (isOverLeft) {
      return LegacyTooltipPosition.BottomLeft
    }
    if (isOverRight) {
      return LegacyTooltipPosition.BottomRight
    }
    return LegacyTooltipPosition.BottomCenter
  }

  if (isOverBottom) {
    if (isOverLeft) {
      return LegacyTooltipPosition.TopLeft
    }
    if (isOverRight) {
      return LegacyTooltipPosition.TopRight
    }
    return LegacyTooltipPosition.TopCenter
  }

  if (isOverLeft) {
    if (isOverTop) {
      return LegacyTooltipPosition.RightTop
    }
    if (isOverBottom) {
      return LegacyTooltipPosition.RightBottom
    }
    return LegacyTooltipPosition.RightCenter
  }

  if (isOverRight) {
    if (isOverTop) {
      return LegacyTooltipPosition.LeftTop
    }
    if (isOverBottom) {
      return LegacyTooltipPosition.LeftBottom
    }
    return LegacyTooltipPosition.LeftCenter
  }

  return placement
}

export function getTooltipStyle({
  tooltipContainer,
  placement,
  offset,
  allowHover,
}: GetTooltipStyle) {
  const {
    top: containerTop,
    left: containerLeft,
    width: containerWidth,
    height: containerHeight,
  } = tooltipContainer.getBoundingClientRect()

  let top = containerTop
  let left = containerLeft
  let translateX = 0
  let translateY = 0

  let offsetTop = 0
  let offsetRight = 0
  let offsetBottom = 0
  let offsetLeft = 0
  let offsetStyle = {}

  switch (placement) {
    case LegacyTooltipPosition.TopCenter:
    case LegacyTooltipPosition.TopLeft:
    case LegacyTooltipPosition.TopRight:
      translateY = -100
      offsetBottom = offset
      break
    case LegacyTooltipPosition.RightCenter:
    case LegacyTooltipPosition.RightTop:
    case LegacyTooltipPosition.RightBottom:
      left += containerWidth
      offsetLeft = offset
      break
    case LegacyTooltipPosition.BottomCenter:
    case LegacyTooltipPosition.BottomLeft:
    case LegacyTooltipPosition.BottomRight:
      top += containerHeight
      offsetTop = offset
      break
    case LegacyTooltipPosition.LeftCenter:
    case LegacyTooltipPosition.LeftTop:
    case LegacyTooltipPosition.LeftBottom:
      translateX = -100
      offsetRight = offset
      break
  }

  switch (placement) {
    case LegacyTooltipPosition.TopCenter:
    case LegacyTooltipPosition.BottomCenter:
      translateX = -50
      left += (containerWidth / 2)
      break
    case LegacyTooltipPosition.TopRight:
    case LegacyTooltipPosition.BottomRight:
      translateX = -100
      left += containerWidth
      break
    case LegacyTooltipPosition.RightCenter:
    case LegacyTooltipPosition.LeftCenter:
      translateY = -50
      top += (containerHeight / 2)
      break
    case LegacyTooltipPosition.RightBottom:
    case LegacyTooltipPosition.LeftBottom:
      translateY = -100
      top += containerHeight
      break
  }

  if (allowHover) {
    offsetStyle = {
      paddingTop: offsetTop,
      paddingRight: offsetRight,
      paddingBottom: offsetBottom,
      paddingLeft: offsetLeft,
    }
  } else {
    top += (offsetTop - offsetBottom)
    left += (offsetLeft - offsetRight)
  }

  return {
    position: 'fixed',
    top,
    left,
    transform: `translate(${translateX}%, ${translateY}%)`,
    ...offsetStyle,
  }
}
