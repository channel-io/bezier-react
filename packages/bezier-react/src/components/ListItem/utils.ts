import { css } from '~/src/foundation'

import { ListItemSize } from './ListItem.types'

export function getStyleOfSize(size?: ListItemSize) {
  switch (size) {
    case ListItemSize.S:
      return css`
        padding: 4px 6px;
        ${({ foundation }) => foundation?.rounding.round6};
      `
    case ListItemSize.L:
      return css`
        padding: 8px 6px;
        ${({ foundation }) => foundation?.rounding.round8};
      `
    case ListItemSize.XL:
      return css`
        padding: 10px 6px;
        ${({ foundation }) => foundation?.rounding.round12};
      `
    case ListItemSize.M:
    default:
      return css`
        padding: 6px;
        ${({ foundation }) => foundation?.rounding.round6};
      `
  }
}
