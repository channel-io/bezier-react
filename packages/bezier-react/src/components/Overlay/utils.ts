/* Internal dependencies */
import { css } from '~/src/foundation'
import {
  OverlayPosition,
  type ContainerRectAttr,
  type TargetRectAttr,
} from './Overlay.types'

const TOP_POSITION_OFFSET = 10

interface GetOverlayPositionArgs {
  containerRect: ContainerRectAttr
  targetRect: TargetRectAttr
  show: boolean
}

export function getOverlayPosition({
  containerRect,
  targetRect,
  show,
}: GetOverlayPositionArgs) {
  const { containerTop, containerLeft, scrollTop, scrollLeft } = containerRect
  const { targetTop, targetLeft, clientTop, clientLeft } = targetRect

  const top = targetTop - clientTop - containerTop + scrollTop
  const left = targetLeft - clientLeft - containerLeft + scrollLeft

  return {
    top: show ? top : top - TOP_POSITION_OFFSET,
    left,
  }
}

interface GetOverlayTranslatationArgs {
  containerRect: ContainerRectAttr
  targetRect: TargetRectAttr | null
  overlay: HTMLElement | null
  position: OverlayPosition
  marginX: number
  marginY: number
  keepInContainer: boolean
}

export function getOverlayTranslation({
  containerRect,
  targetRect,
  overlay,
  position,
  marginX,
  marginY,
  keepInContainer,
}: GetOverlayTranslatationArgs) {
  if (containerRect && targetRect && overlay) {
    const {
      containerWidth,
      containerHeight,
      containerTop,
      containerLeft,
    } = containerRect
    const { targetWidth, targetHeight, targetTop, targetLeft } = targetRect
    const { width: overlayWidth, height: overlayHeight } = overlay.getBoundingClientRect()

    let translateX = 0
    let translateY = 0

    // pre position
    switch (position) {
      case OverlayPosition.TopCenter:
      case OverlayPosition.TopLeft:
      case OverlayPosition.TopRight:
        translateY -= (overlayHeight + marginY)
        translateX += marginX
        break
      case OverlayPosition.RightCenter:
      case OverlayPosition.RightTop:
      case OverlayPosition.RightBottom:
        translateX += (targetWidth + marginX)
        translateY += marginY
        break
      case OverlayPosition.BottomCenter:
      case OverlayPosition.BottomLeft:
      case OverlayPosition.BottomRight:
        translateY += (targetHeight + marginY)
        translateX += marginX
        break
      case OverlayPosition.LeftCenter:
      case OverlayPosition.LeftTop:
      case OverlayPosition.LeftBottom:
        translateX -= (overlayWidth + marginX)
        translateY += marginY
        break
    }

    switch (position) {
      // post position
      case OverlayPosition.TopCenter:
      case OverlayPosition.BottomCenter:
        translateX -= ((overlayWidth / 2) - (targetWidth / 2))
        break
      case OverlayPosition.TopRight:
      case OverlayPosition.BottomRight:
        translateX -= (overlayWidth - targetWidth)
        break
      case OverlayPosition.RightCenter:
      case OverlayPosition.LeftCenter:
        translateY -= ((overlayHeight / 2) - (targetHeight / 2))
        break
      case OverlayPosition.RightBottom:
      case OverlayPosition.LeftBottom:
        translateY -= (overlayHeight - targetHeight)
        break

      // inner position
      case OverlayPosition.InnerLeftTop:
        translateX = marginX
        translateY = marginY
        break
      case OverlayPosition.InnerLeftBottom:
        translateX = marginX
        translateY = targetHeight - overlayHeight + marginY
        break
      case OverlayPosition.InnerRightTop:
        translateX = targetWidth - overlayWidth + marginX
        translateY = marginY
        break
      case OverlayPosition.InnerRightBottom:
        translateX = targetWidth - overlayWidth + marginX
        translateY = targetHeight - overlayHeight + marginY
        break
    }

    if (keepInContainer) {
      const isOverTop = targetTop + translateY < containerTop
      const isOverBottom = targetTop + translateY + overlayHeight > containerTop + containerHeight
      const isOverLeft = targetLeft + translateX < containerLeft
      const isOverRight = targetLeft + translateX + overlayWidth > containerLeft + containerWidth

      const topSpace = targetTop - containerTop
      const bottomSpace = (containerTop + containerHeight) - (targetTop + targetHeight)
      const leftSpace = targetLeft - containerLeft
      const rightSpace = (containerLeft + containerWidth) - (targetLeft + targetWidth)

      if (isOverTop) {
        translateY = topSpace > bottomSpace
          ? translateY
          : targetHeight - translateY - overlayHeight
      } else if (isOverBottom) {
        translateY = bottomSpace > topSpace
          ? translateY
          : targetHeight - translateY - overlayHeight
      }
      if (isOverLeft) {
        translateX = leftSpace > rightSpace
          ? translateX
          : targetWidth - translateX - overlayWidth
      } else if (isOverRight) {
        translateX = rightSpace > leftSpace
          ? translateX
          : targetWidth - translateX - overlayWidth
      }
    }

    return {
      translateX: Math.round(translateX),
      translateY: Math.round(translateY),
    }
  }
  return {
    translateX: 0,
    translateY: 0,
  }
}

interface GetOverlayStyleArgs {
  containerRect: ContainerRectAttr
  targetRect: TargetRectAttr | null
  overlay: HTMLElement | null
  position: OverlayPosition
  marginX: number
  marginY: number
  keepInContainer: boolean
  show: boolean
}

export function getOverlayStyle({
  containerRect,
  targetRect,
  overlay,
  position,
  marginX,
  marginY,
  keepInContainer,
  show,
}: GetOverlayStyleArgs) {
  if (containerRect && targetRect) {
    const { top, left } = getOverlayPosition({ containerRect, targetRect, show })
    const { translateX, translateY } = getOverlayTranslation({
      containerRect,
      targetRect,
      overlay,
      position,
      marginX,
      marginY,
      keepInContainer,
    })

    return css`
      top: ${top}px;
      left: ${left}px;
      transform: translateX(${translateX}px) translateY(${translateY}px);
    `
  }
  return css``
}
