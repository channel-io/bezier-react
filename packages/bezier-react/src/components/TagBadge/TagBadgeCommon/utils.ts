import { type SemanticColor } from '~/src/types/tokens'

import TagBadgeSize from './constants/TagBadgeSize'
import type TagBadgeVariant from './constants/TagBadgeVariant'
import { TagBadgeBgColorPreset } from './constants/TagBadgeVariant'

export function getProperTagBadgeTypo(size: TagBadgeSize) {
  return {
    [TagBadgeSize.XS]: '11' as const,
    [TagBadgeSize.S]: '13' as const,
    [TagBadgeSize.M]: '14' as const,
    [TagBadgeSize.L]: '15' as const,
  }[size]
}

export function getProperTagBadgeBgColor(color: TagBadgeVariant): SemanticColor {
  return TagBadgeBgColorPreset[color]
}
