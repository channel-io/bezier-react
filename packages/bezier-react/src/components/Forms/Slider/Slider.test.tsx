import React from 'react'

import {
  isInaccessible,
  waitFor,
} from '@testing-library/react'
import userEvent from '@testing-library/user-event'

import { render } from '~/src/utils/test'

import {
  SLIDER_TEST_ID,
  Slider,
} from './Slider'
import type SliderProps from './Slider.types'

describe('Slider', () => {
  const renderSlider = (props?: Partial<SliderProps>) => render(<Slider {...props} />)

  let user: ReturnType<typeof userEvent.setup>

  beforeEach(() => {
    user = userEvent.setup()
  })

  describe('Snapshot', () => {
    it('should match snapshot', () => {
      const { container } = renderSlider({ defaultValue: [5], guide: [5] })
      expect(container).toMatchSnapshot()
    })
  })

  describe('ARIA', () => {
    it('should be accessible', () => {
      const { container } = renderSlider()
      expect(isInaccessible(container)).toBe(false)
    })

    describe('role', () => {
      it('should render thumb with "slider" role', () => {
        const { getAllByRole } = renderSlider()
        const sliderThumb = getAllByRole('slider')
        expect(sliderThumb).toHaveLength(1)
        expect(sliderThumb[0]).toBeInTheDocument()
      })

      it('should render multiple thumb when multiple defaultValue is specified', () => {
        const { getAllByRole } = renderSlider({ defaultValue: [4, 5, 6] })
        const sliderThumb = getAllByRole('slider')
        expect(sliderThumb).toHaveLength(3)
        expect(sliderThumb[0]).toBeInTheDocument()
        expect(sliderThumb[1]).toBeInTheDocument()
        expect(sliderThumb[2]).toBeInTheDocument()
      })

      it('should not render thumb when defaultValue is not specified', () => {
        const { queryAllByRole } = renderSlider({ defaultValue: [] })
        const sliderThumb = queryAllByRole('slider')
        expect(sliderThumb).toHaveLength(0)
      })
    })

    describe('aria-valuemax, aria-valuemin', () => {
      it('should have default value on "aria-valuemin", "aria-valuemax" attribute when no props specified', () => {
        const { getAllByRole } = renderSlider()
        const sliderThumb = getAllByRole('slider')
        expect(sliderThumb).toHaveLength(1)
        expect(sliderThumb[0]).toHaveAttribute('aria-valuemin', '0')
        expect(sliderThumb[0]).toHaveAttribute('aria-valuemax', '10')
      })

      it('should have proper value on "aria-valuemin", "aria-valuemax" attribute', () => {
        const { getAllByRole } = renderSlider({ min: -50, max: 70 })
        const sliderThumb = getAllByRole('slider')
        expect(sliderThumb).toHaveLength(1)
        expect(sliderThumb[0]).toHaveAttribute('aria-valuemin', '-50')
        expect(sliderThumb[0]).toHaveAttribute('aria-valuemax', '70')
      })
    })

    describe('aria-valuenow', () => {
      it('should have default value on "aria-valuenow" attribute when no props specified', () => {
        const { getAllByRole } = renderSlider()
        const sliderThumb = getAllByRole('slider')

        expect(sliderThumb).toHaveLength(1)
        expect(sliderThumb[0]).toHaveAttribute('aria-valuenow', '0')
      })

      it('should have proper value on "aria-valuenow" attribute', () => {
        const { getAllByRole } = renderSlider({ defaultValue: [2, 6, 7] })
        const sliderThumb = getAllByRole('slider')

        expect(sliderThumb).toHaveLength(3)
        expect(sliderThumb[0]).toHaveAttribute('aria-valuenow', '2')
        expect(sliderThumb[1]).toHaveAttribute('aria-valuenow', '6')
        expect(sliderThumb[2]).toHaveAttribute('aria-valuenow', '7')
      })
    })

    describe('aria-disabled', () => {
      it('should have "false" value on "aria-disabled" attribute when disabled prop is false', () => {
        const { getByTestId } = renderSlider({ disabled: false })
        const slider = getByTestId(SLIDER_TEST_ID)
        expect(slider).toHaveAttribute('aria-disabled', 'false')
      })

      it('should have "true" value on "aria-disabled" attribute when disabled prop is true', () => {
        const { getByTestId } = renderSlider({ disabled: true })
        const slider = getByTestId(SLIDER_TEST_ID)
        expect(slider).toHaveAttribute('aria-disabled', 'true')
      })
    })
  })

  describe('User Interactions', () => {
    it('should focus on the thumb when clicked', async () => {
      const { getByRole } = renderSlider()
      const sliderThumb = getByRole('slider')
      await user.click(sliderThumb)
      expect(sliderThumb).toHaveFocus()
    })

    it('should focus on the thumb when user presses tab key', async () => {
      const { getByRole } = renderSlider()
      const sliderThumb = getByRole('slider')
      await user.tab()
      expect(sliderThumb).toHaveFocus()
    })

    it('should not focus on the thumb when disabled', async () => {
      const { getByRole } = renderSlider({ disabled: true })
      const sliderThumb = getByRole('slider')
      await user.click(sliderThumb)
      expect(sliderThumb).not.toHaveFocus()
      await user.tab()
      expect(sliderThumb).not.toHaveFocus()
    })

    it('should increments/decrements by the step value when user presses Arrow key', async () => {
      const { getByRole } = renderSlider({ defaultValue: [0], step: 2 })
      const sliderThumb = getByRole('slider')
      await user.tab()
      await user.keyboard('[ArrowRight]')
      expect(sliderThumb).toHaveAttribute('aria-valuenow', '2')
      await user.keyboard('[ArrowLeft]')
      expect(sliderThumb).toHaveAttribute('aria-valuenow', '0')
      await user.keyboard('[ArrowUp]')
      expect(sliderThumb).toHaveAttribute('aria-valuenow', '2')
      await user.keyboard('[ArrowDown]')
      expect(sliderThumb).toHaveAttribute('aria-valuenow', '0')
    })

    it('should increments/decrements by the larger step value when user presses PageUp/PageDown key', async () => {
      const { getByRole } = renderSlider({ defaultValue: [0] })
      const sliderThumb = getByRole('slider')
      await user.tab()
      await user.keyboard('[PageUp]')
      expect(sliderThumb).toHaveAttribute('aria-valuenow', '10')
      await user.keyboard('[PageDown]')
      expect(sliderThumb).toHaveAttribute('aria-valuenow', '0')
    })

    it('should increments/decrements by the larger step value when user presses Shift + ArrowUp/Down key', async () => {
      const { getByRole } = renderSlider({ defaultValue: [0] })
      const sliderThumb = getByRole('slider')
      await user.tab()
      await user.keyboard('{Shift>}[ArrowUp]')
      expect(sliderThumb).toHaveAttribute('aria-valuenow', '10')
      await user.keyboard('{Shift>}[ArrowDown]')
      expect(sliderThumb).toHaveAttribute('aria-valuenow', '0')
    })

    it('should set the maximum/minimum value when user presses Home/End key', async () => {
      const { getByRole } = renderSlider({ defaultValue: [0], min: 10, max: 100 })
      const sliderThumb = getByRole('slider')
      await user.tab()
      await user.keyboard('[End]')
      expect(sliderThumb).toHaveAttribute('aria-valuenow', '100')
      await user.keyboard('[Home]')
      expect(sliderThumb).toHaveAttribute('aria-valuenow', '10')
    })

    it('`onValueChange` function should be executed when the value changes', async () => {
      const onValueChange = jest.fn()
      renderSlider({ onValueChange })
      await user.tab()
      await user.keyboard('[ArrowRight]')
      expect(onValueChange).toHaveBeenCalledTimes(1)
    })

    it('If there are more than one thumb, the steps between them should not be less than `minStepsBetweenThumbs`.', async () => {
      const { getAllByRole } = renderSlider({ defaultValue: [0, 2], minStepsBetweenThumbs: 2 })
      const [firstThumb, secondThumb] = getAllByRole('slider')
      await user.tab()
      await user.keyboard('[ArrowRight]')
      expect(firstThumb).not.toHaveAttribute('aria-valuenow', '1')
      await user.tab()
      await user.keyboard('[ArrowLeft]')
      expect(secondThumb).not.toHaveAttribute('aria-valuenow', '1')
    })
  })

  describe('Style', () => {
    it('should render Slider with given width (number)', () => {
      const { getByTestId } = renderSlider({ width: 300 })
      const slider = getByTestId(SLIDER_TEST_ID)
      expect(slider).toHaveStyle('--b-slider-width: 300px')
    })

    it('should render Slider with given width (string)', () => {
      const { getByTestId } = renderSlider({ width: '500px' })
      const slider = getByTestId(SLIDER_TEST_ID)
      expect(slider).toHaveStyle('--b-slider-width: 500px')
    })
  })

  describe('Tooltip', () => {
    it('should show tooltip when user hovers on the thumb', async () => {
      const { getByRole, getAllByText } = renderSlider({ defaultValue: [0] })
      const sliderThumb = getByRole('slider')
      await user.hover(sliderThumb)
      await waitFor(() => {
        expect(getAllByText('0')[0]).toBeInTheDocument()
      }, { timeout: 10000 })
    })

    it('should show tooltip when user focuses on the thumb', async () => {
      const { getAllByText } = renderSlider({ defaultValue: [0] })
      await user.tab()
      expect(getAllByText('0')[0]).toBeInTheDocument()
    })

    it('should show tooltip when user clicks on the thumb', async () => {
      const { getByRole, getAllByText } = renderSlider({ defaultValue: [0] })
      const sliderThumb = getByRole('slider')
      await user.click(sliderThumb)
      await waitFor(() => {
        expect(getByRole('tooltip')).toBeInTheDocument()
        expect(getAllByText('0')[0]).toBeInTheDocument()
      }, { timeout: 10000 })
    })

    it('should not show tooltip when disableTooltip is true', async () => {
      const { getByRole, queryByRole, queryAllByText } = renderSlider({ defaultValue: [0], disableTooltip: true })
      const sliderThumb = getByRole('slider')
      await user.click(sliderThumb)
      expect(queryByRole('tooltip')).not.toBeInTheDocument()
      expect(queryAllByText('0')[0]).toBeUndefined()
    })
  })
})
