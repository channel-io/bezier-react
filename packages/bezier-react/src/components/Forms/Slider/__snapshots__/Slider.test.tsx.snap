// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slider Snapshot should match snapshot 1`] = `
.c0 {
  --bezier-slider-width: auto;
  --bezier-slider-thumb-size: 20px;
  position: relative;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  width: var(--bezier-slider-width);
  height: var(--bezier-slider-thumb-size);
  touch-action: none;
  cursor: pointer;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: opacity;
  transition-property: opacity;
}

.c0[data-disabled] {
  cursor: not-allowed;
  opacity: 0.4;
}

.c1 {
  position: relative;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  height: 6px;
  background-color: var(--bg-black-dark);
  border-radius: 3px;
}

.c2 {
  position: absolute;
  height: 100%;
  background-color: var(--bgtxt-green-normal);
  border-radius: 3px;
}

.c3 {
  --bezier-slider-guide-height: 8px;
  position: absolute;
  top: calc(-1 * (var(--bezier-slider-guide-height) + 3px));
  left: calc(var(--bezier-slider-thumb-size) / 2);
  width: calc(100% - var(--bezier-slider-thumb-size));
}

.c4 {
  --bezier-slider-guide-left: 0;
  position: absolute;
  top: 0;
  left: var(--bezier-slider-guide-left);
  width: 2px;
  height: var(--bezier-slider-guide-height);
  background-color: var(--bg-black-light);
  border-radius: 1px;
  -webkit-transform: translateX(-50%);
  -ms-transform: translateX(-50%);
  transform: translateX(-50%);
}

.c5 {
  all: unset;
  display: block;
  width: var(--bezier-slider-thumb-size);
  height: var(--bezier-slider-thumb-size);
  border-radius: 12px;
  background-color: #FFFFFF;
  box-shadow: inset 0 0 2px 0 #FFFFFF1F, 0 0 2px 1px #0000000D, 0 2px 6px #00000014;
  background-color: var(--bgtxt-absolute-white-dark);
  -webkit-transition-delay: 0ms;
  transition-delay: 0ms;
  -webkit-transition-timing-function: cubic-bezier(.3,0,0,1);
  transition-timing-function: cubic-bezier(.3,0,0,1);
  -webkit-transition-duration: 150ms;
  transition-duration: 150ms;
  -webkit-transition-property: box-shadow;
  transition-property: box-shadow;
}

.c5:hover {
  background-color: #FFFFFF;
  box-shadow: inset 0 0 2px 0 #FFFFFF1F, 0 0 2px 1px #0000000D, 0 4px 12px #00000026;
  background-color: var(--bgtxt-absolute-white-dark);
}

.c5:focus-visible {
  box-shadow: 0 0 0 3px #5E56F04D, inset 0 0 0 1px #5E56F0;
}

<div>
  <span
    aria-disabled="false"
    class="c0"
    data-orientation="horizontal"
    data-testid="bezier-react-slider"
    dir="ltr"
    foundation="[object Object]"
    style="--bezier-slider-width: 36px; --radix-slider-thumb-transform: translateX(-50%);"
  >
    <span
      class="c1"
      data-orientation="horizontal"
      foundation="[object Object]"
    >
      <span
        class="c2"
        data-orientation="horizontal"
        foundation="[object Object]"
        style="left: 0%; right: 50%;"
      />
      <div
        class="c3"
      >
        <div
          class="c4"
          style="--bezier-slider-guide-left: 50%;"
        />
      </div>
    </span>
    <span
      style="transform: var(--radix-slider-thumb-transform); position: absolute; left: calc(50% + 0px);"
    >
      <div
        aria-orientation="horizontal"
        aria-valuemax="10"
        aria-valuemin="0"
        aria-valuenow="5"
        class="c5"
        data-orientation="horizontal"
        data-radix-collection-item=""
        data-state="closed"
        role="slider"
        tabindex="0"
      />
    </span>
  </span>
</div>
`;
